### PR TITLE
feat: Implement function to calcualte charge cost

### PR DESCRIPTION
**Pull request by CodeBear801**
issue: https://github.com/Telenav/osrm-backend/issues/143

- Calculate charging cost for given status and target status.  
  + `ChargeStrategy` creates several charge solutions via function `CreateChargingStrategies`, which means charge battery to a certain certain energy level.
  + When user reaches a charge station, he has a inital arrival energy and target status, he could use `EvaluateCost` to calculate cost needed, such as time, money, etc.
- Add NullChargeStrategy for unit tests
- Fix the issue of hard coded `chargeTime`
- Remove test cases for chargeEnergy, which will be added later in the future pull

## Tasklist

 - [x] modify code
 - [x] add tests 
 - [ ] review

### CODE REVIEW

**Comment by CodeBear801** [comment]
This pull request is target for:

- Calculate charging cost for given status and target status.  
  + `ChargeStrategy` creates several charge solutions via function `CreateChargingStrategies`, which means charge battery to a certain certain energy level.
  + When user reaches a charge station, he has a inital arrival energy and target status, he could use `EvaluateCost` to calculate cost needed, such as time, money, etc.
- Add NullChargeStrategy for unit tests
- Fix the issue of hard coded `chargeTime`
- Remove test cases for chargeEnergy, which will be added later in the future pull

To be done later:
- Change interface name from `ChargingStrategyCreator` to `ChargingStrategy`
- Change interface func from `CreateChargingStrategies` to `CreateChargingSolutions`

File: integration/oasis/chargingstrategy/chargingstrategy.go [hunk 1]
```diff
@@ -11,24 +15,85 @@ func NewFakeChargingStrategyCreator(maxEnergyLevel float64) *fakeChargingStrateg
...
+		}
+	}
+
+	glog.Fatalf("Invalid charging state %#v\n", targetState)
+	return noNeedCharge
+}
+
+var epsilon float64 = 0.00000001
+
+func floatEquals(a, b float64) bool {
```
Comment (wangyoucao577) [inline integration/oasis/chargingstrategy/chargingstrategy.go]: We really need to find/implement a `floatEquals` package. There're too many places want to use it.

File: integration/oasis/chargingstrategy/charge_strategy_creator.go [hunk 2]
```diff
@@ -1,14 +1,22 @@
...
-// ChargingStrategy contains charging related information
-type ChargingStrategy struct {
-	ChargingTime   float64
+// ChargingStatus contains charging related information
+type ChargingStatus struct {
 	ChargingEnergy float64
 }
 
+// ChargingCost represents the cost needed to reach certain states
+type ChargingCost struct {
```
Comment (wangyoucao577) [inline integration/oasis/chargingstrategy/charge_strategy_creator.go]: Maybe `ChargingStatus`->`State` and `CharingCost`->`Cost` will be better? See https://github.com/golang/go/wiki/CodeReviewComments#package-names

File: integration/oasis/chargingstrategy/chargingstrategy.go [hunk 3]
```diff
@@ -11,24 +15,85 @@ func NewFakeChargingStrategyCreator(maxEnergyLevel float64) *fakeChargingStrateg
...
 		},
 	}
 }
+
+// Fake charge strategy
+// From empty energy:
+//                    1 hour charge to 60% of max energy
+//                    2 hour charge to 80%, means from 60% ~ 80% need 1 hour
+//                    4 hour charge to 100%, means from 80% ~ 100% need 2 hours
+func (f *fakeChargingStrategyCreator) EvaluateCost(arrivalEnergy float64, targetState ChargingStatus) ChargingCost {
```
Comment (wangyoucao577) [inline integration/oasis/chargingstrategy/chargingstrategy.go]: `targetState` or `targetStatus`, which one would be better?

File: integration/oasis/chargingstrategy/null_charge_strategy.go [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package chargingstrategy
+
+type nullChargeStrategy struct {
```
Comment (wangyoucao577) [inline integration/oasis/chargingstrategy/null_charge_strategy.go]: same issue for naming. see https://github.com/golang/go/wiki/CodeReviewComments#package-names

File: integration/oasis/chargingstrategy/chargingstrategy.go (hunk 1 above)
Comment (CodeBear801) [reply integration/oasis/chargingstrategy/chargingstrategy.go]: Yes, and `location`

Comment (CodeBear801) [reply integration/oasis/chargingstrategy/chargingstrategy.go]: How about add `common` folder in `integration`
- For floatEquals, we add into `util.go`.  Search this from other repository, one sample is [`go-gl/mathgl/mathgl32/util`](https://github.com/go-gl/mathgl/blob/5ab0e04e1f55a10f3a71b1fccc60ca2ee9a0bc02/mgl32/util.go#L40:6)
- For `Location` could put which into `type` folder

File: integration/oasis/chargingstrategy/charge_strategy_creator.go (hunk 2 above)
Comment (CodeBear801) [reply integration/oasis/chargingstrategy/charge_strategy_creator.go]: Will do it in future pull: https://github.com/Telenav/osrm-backend/issues/211

File: integration/oasis/chargingstrategy/chargingstrategy.go (hunk 3 above)
Comment (CodeBear801) [reply integration/oasis/chargingstrategy/chargingstrategy.go]: Keep them consistency is most important, will keep `targetState`

File: integration/oasis/chargingstrategy/chargingstrategy.go (hunk 1 above)
Comment (wangyoucao577) [reply integration/oasis/chargingstrategy/chargingstrategy.go]: Not only `oasis` need it, but also other packages. What if create a `integration/util` package for the `floatEquals`, and  a `integration/map` package for the `Location`? The `Location` here is for general purpose, if some special usage(e.g., `Coordinate` in `osrm`) just create their owns.

Comment (CodeBear801) [reply integration/oasis/chargingstrategy/chargingstrategy.go]: Ok, will put it to `integration`

**Pull request merged**